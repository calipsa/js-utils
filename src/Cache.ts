import Singleton from './Singleton'

export default class Cache<K, V> extends Singleton {
  map = new Map<K, V>()

  add(idx: K, item: V) {
    this.map.set(idx, item)
  }

  get(idx: K) {
    return this.map.get(idx)
  }

  clear() {
    this.map.clear()
  }
}
