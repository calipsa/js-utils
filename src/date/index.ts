import fromNow from './fromNow'

export {
  // eslint-disable-next-line import/prefer-default-export
  fromNow,
}
