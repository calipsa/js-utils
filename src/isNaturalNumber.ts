/**
 * Checks if a value is a natural number,
 * i.e. is an integer greater than zero
 */
export default (num: any): num is number =>
  Number.isInteger(num) && num > 0
