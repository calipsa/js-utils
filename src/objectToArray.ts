import isNonNegativeInteger from './isNonNegativeInteger'

type O = Readonly<{
  [k: number]: any,
}>

export default <Obj extends O, V extends Obj[keyof Obj]>(o: Obj) => {
  const indices = Object.keys(o)
    .map(Number)
    .filter(isNonNegativeInteger)

  const minKey = Math.min(...indices)
  const maxKey = Math.max(...indices)
  const arr: V[] = []

  for (let i = minKey; i <= maxKey; ++i) {
    arr[i] = o[i]
  }

  return arr
}
