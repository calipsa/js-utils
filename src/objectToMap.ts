/**
 * Converts an object to a `Map`
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export default <T extends {}, K extends keyof T>(o: T) =>
  new Map<string, T[K]>(Object.entries(o))
