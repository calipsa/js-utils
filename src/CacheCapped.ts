import Cache from './Cache'

export default class CacheCapped<K, V> extends Cache<K, V> {
  private capacity: number

  constructor(capacity: number) {
    super()
    this.capacity = capacity
  }

  rm(item: K) {
    this.map.delete(item)
  }

  add(idx: K, item: V) {
    if (this.map.size === this.capacity && !this.map.has(idx)) {
      const oldest = this.map.keys().next().value
      this.rm(oldest)
    }
    this.map.set(idx, item)
  }
}
