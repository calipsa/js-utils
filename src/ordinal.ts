import isNaturalNumber from './isNaturalNumber'

const got = (i: any) =>
  typeof i === 'string' ? JSON.stringify(i) : i

export default (i: number) => {
  if (!isNaturalNumber(i)) {
    throw new Error(`Number must be a natural number. Instead got: ${got(i)}`)
  }

  const j = i % 10
  const k = (i % 100) - j
  if (k !== 10) {
    if (j === 1) {
      return `${i}st`
    }
    if (j === 2) {
      return `${i}nd`
    }
    if (j === 3) {
      return `${i}rd`
    }
  }
  return `${i}th`
}
