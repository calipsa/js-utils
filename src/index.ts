import Cache from './Cache'
import CacheCapped from './CacheCapped'
import delay from './delay'
import hasDuplicates from './hasDuplicates'
import isNonNegativeInteger from './isNonNegativeInteger'
import objectToArray from './objectToArray'
import objectToFunction from './objectToFunction'
import objectToFunctionWithBind from './objectToFunctionWithBind'
import Singleton from './Singleton'
import sliceRotate from './sliceRotate'
import toNearestByStep from './toNearestByStep'

export {
  Cache,
  CacheCapped,
  delay,
  hasDuplicates,
  isNonNegativeInteger,
  objectToArray,
  objectToFunction,
  objectToFunctionWithBind,
  Singleton,
  sliceRotate,
  toNearestByStep,
}
