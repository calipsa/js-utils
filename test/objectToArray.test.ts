import { objectToArray } from '../src'

describe('objectToArray', () => {
  it('Should convert object to array', () => {
    const o = {
      3: 'd',
      '10': 'k',
      '0': 'a',
      '1': 'b',
      '6': 'g',
      8: 'i',
      11: 'l',
      '4': 'e',
      7: 'h',
      13: 'n',
      '9': 'j',
      '12': 'm',
      '2': 'c',
      5: 'f',
    }
    const result = objectToArray(o)
    expect(result).toMatchSnapshot()
  })
})
