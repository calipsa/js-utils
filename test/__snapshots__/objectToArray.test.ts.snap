// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`objectToArray Should convert object to array 1`] = `
Array [
  "a",
  "b",
  "c",
  "d",
  "e",
  "f",
  "g",
  "h",
  "i",
  "j",
  "k",
  "l",
  "m",
  "n",
]
`;
